#!/bin/bash

outdir="hybrid_RNAseq"
rawdata_dir="raw_data"
genome_star_index_dir="star_index"
gtf_file="merge.XENTR_10.0_XENLA_9.2_GCF.gtf"
rsem_reference_name="merge.XENTR_10.0_XENLA_9.2"

for sample in ${rawdata_dir}/*1.fq.gz;do
	sample=$(basename $sample _R1.fq.gz)
	R1=$rawdata_dir/${sample}_R1.fq.gz
	R2=$rawdata_dir/${sample}_R2.fq.gz
	cleanR1=${outdir}/clean_data/${sample}.clean_1P.fastq.gz
	cleanR2=${outdir}/clean_data/${sample}.clean_2P.fastq.gz

	if [ ! -d $outdir/bam ];then
		mkdir -p ${outdir}/bam
	fi
	bamdir=${outdir}/bam

	if [ ! -d $outdir/rsem ];then
		mkdir -p ${outdir}/rsem
	fi
	rsemdir=${outdir}/rsem

	echo -e "$(date '+%Y-%m-%d %H:%M:%S') $sample processing ... "

## 1. do qc for raw data (raw data refers to cleandata generated by sequencing).
	if [ ! -d $outdir/fastqc/clean_qc1 ];then
		mkdir -p $outdir/fastqc/clean_qc1
	fi

	fastqc -t 5 -o $outdir/fastqc/clean_qc1 $rawdata_dir/${sample}_R1.fq.gz
	fastqc -t 5 -o $outdir/fastqc/clean_qc1 $rawdata_dir/${sample}_R2.fq.gz
	echo -e "$(date '+%Y-%m-%d %H:%M:%S') $sample fastqc raw reads ..."

## 2. trim to remove low-quality sequences.
	if [ ! -d $outdir/cleandata ];then
		mkdir -p $outdir/cleandata
	fi

    java -jar /work/bio-xujs/biosoft/Trimmomatic/Trimmomatic-0.38/trimmomatic-0.38.jar PE -phred33 \
        -threads 10 $R1 $R2 \
        -baseout $outdir/clean_data/${sample}.clean.fastq.gz \
        ILLUMINACLIP:/work/bio-xujs/biosoft/Trimmomatic/Trimmomatic-0.38/adapters/TruSeq3-PE.fa:2:30:10 \
                        HEADCROP:15 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:75
	wait
	echo -e "$(date '+%Y-%m-%d %H:%M:%S') $sample Trimming ..."

## 3. do qc align to check quality.
	if [ ! -d $outdir/fastqc/qc2 ];then
		mkdir -p $outdir/fastqc/qc2
	fi

	fastqc -t 10 -o $outdir/fastqc/clean_qc2 ${cleanR1}
	fastqc -t 10 -o $outdir/fastqc/clean_qc2 ${cleanR2}
	wait
	echo -e "$(date '+%Y-%m-%d %H:%M:%S') $sample fastqc clean reads ..."

## 4. align the reads to the reference: STAR.
## To prevent misalignment to other species, apply stringent criteria:
##				—— --outSAMmultNmax 1 --outFilterMismatchNmax 0 --outSAMmapqUnique 10
	cd ${bamdir}
	STAR --twopassMode Basic --quantMode TranscriptomeSAM GeneCounts \
		--sjdbGTFfile $gtf_file \
		--runMode alignReads \
		--runThreadN 15 --genomeDir ${genome_star_index_dir} \
		--outSAMtype BAM SortedByCoordinate \
		--outSAMmultNmax 1 --outFileNamePrefix ${sample} \
		--outSAMmapqUnique 10 --readFilesCommand zcat \
		--outFilterMismatchNmax 0 \
		--readFilesIn ${cleanR1} ${cleanR2}

## 5. quantitative analysis: RSEM.
	cd ${rsemdir}
	rsem-calculate-expression --paired-end -p 15 -no-bam-output --alignments ${bamdir}/${sample}Aligned.toTranscriptome.out.bam ${rsem_reference_name} ${sample}

done